torch/storage.py

```diff
@@ -163,3 +139,288 @@ def _load_from_bytes(b):
+                assert isinstance(dtype, torch.dtype)
+                self.dtype = dtype
+
+            else:
+                assert hasattr(self, 'dtype')
+                assert len(kwargs) == 1, (
+                    f"Only 'wrap_storage' should be given, but got: {kwargs.keys()}")
+                dtype = self.dtype
+
+            storage = kwargs['wrap_storage']
+
+            # TODO: change `self.storage` to `self._storage`
+            if isinstance(storage, TypedStorage):
+                # TODO: Consider whether we should actually allow this or not.
+                # I think it might be useful for .float(), .double(), etc
+                # methods?
+                self.storage = storage.storage
+
+            else:
+                assert isinstance(storage, eval(storage.__module__ + ".ByteStorage"))
+                self.storage = storage
+        else:
+            assert type(self) != TypedStorage, (
+                "Calling __init__ this way is only supported in TypedStorage's "
+                "child classes. TypedStorage can only be directly instantiated "
+                "when kwargs 'wrap_storage' and 'dtype' are given.")
+
+            # TODO: Still need to override some of the constructor behavior.
+            # For instance, when we do something like
+            # `torch.FloatStorage([0, 1, 2])`, we'll end up calling
+            # `torch.ByteStorage([0, 1, 2])` here, which is not correct,
+            # since we want the storage to be filled with 3 float values
+            # [0, 1, 2]
+            self.storage = eval(self.__module__ + ".ByteStorage")(*args, **kwargs)
+
+    def _untyped(self):
+        return self.storage
+
+    def _new_wrapped_storage(self, byte_storage):
+        module = eval(byte_storage.__module__)
+        assert type(byte_storage) == module.ByteStorage
+
+        if type(self) == TypedStorage:
+            return TypedStorage(wrap_storage=byte_storage, dtype=self.dtype)
+        else:
+            # NOTE: We need to use the module of byte_storage in case self's
+            # module is different. e.g. if self is on CPU and byte_storage
+            # is on CUDA, and vice versa
+            return getattr(module, type(self).__name__)(wrap_storage=byte_storage)
+
+    def __len__(self):
+        return self.storage.nbytes() // self.element_size()
+
+    def _maybe_wrap_index(self, idx):
+        if type(idx) != int:
+            raise TypeError(
+                f"can't index a {type(self)} with {type(idx)}")
+
+        if idx is None:
+            return 0
+
+        else:
+            if (idx >= self.size()) or (idx < -self.size()):
+                raise IndexError(
+                    f'index {idx} out of range for storage of size {self.size()}')
+            return idx % self.size()
+
+
+    def __setitem__(self, idx, value):
+        # TODO: Seems bad to use a tensor inside a storage method
+        tmp_tensor = torch.tensor(self.storage, device=self.device, dtype=self.dtype)
```

Comment:
This doesn't seem like a very good way to implement this (and other similar funcs, like in `__getitem__`), but I'm not really sure how else to do it.

It should also be possible to convert `value` to a byte array and set the underlying ByteStorage with it, byte-by-byte. But converting it to the correct byte array would be complicated, since `value` needs to be cast to the storage's dtype first.